### ISSUE TITLE
block-scoped-var fails on mutual recursion

### ISSUE THREAD

**Issue by RReverser** [eslint/eslint#2747]
Simplified my case to the following code:

```javascript
function a(n) {
	return n > 0 ? b(n - 1) : "a";
}

function b(n) {
	return n > 0 ? a(n - 1) : "b";
}
```

And simplified eslintrc that allows to reproduce the issue:

```json
{
  "ecmaFeatures": {
    "blockBindings": true,
    "modules": true
  },
  "rules": {
    "no-use-before-define": 2,
    "block-scoped-var": 2
  }
}
```

Without `blockBindings` enabled, only `no-use-before-define` fails.
Turning `modules` on/off doesn't affect the result, only adds `use strict`-related check in this case.

**Comment by michaelficarra** [comment]
@RReverser `no-use-before-define` doesn't allow the reference to appear anywhere before the definition in the program text. The more sane configuration, in my opinion, is the `nofunc` option, and it should fix your problem.

**Comment by RReverser** [comment]
@michaelficarra Doesn't help for classes though (which was my original case).

I'd just wish the rule would be somewhat smarter and wouldn't warn when definition *outside* of the function/class boundary is referenced.

Also - what about `block-scoped-var`?

**Comment by michaelficarra** [comment]
There's a proposal to fix that: https://github.com/eslint/eslint/issues/2750

What's the problem you're having with `block-scoped-var`?

**Comment by RReverser** [comment]
See title - it also fails with given example & config.

**Comment by RReverser** [comment]
> There's a proposal to fix that: #2750

Haha, just in time :)

**Comment by nzakas** [comment]
@RReverser can you provide the actual ESLint output as well as what you'd expect the output to be?

**Comment by RReverser** [comment]
Sure, here it is:
```
x.js
  2:19  error  b was used before it was defined     no-use-before-define
  2:19  error  "b" used outside of binding context  block-scoped-var

✖ 2 problems (2 errors, 0 warnings)
```

I expected not to see any of these errors.

**Comment by nzakas** [comment]
It looks like nofunc solves your problem here, and #2750 would take care of the rest of your concerns there. So let's focus this issue on `block-scoped-var`.

`block-scoped-var` probably would need to be rewritten in order to properly support hoisting, as this looks like a bug.

**Comment by RReverser** [comment]
Thanks!

**Comment by mysticatea** [comment]
I will try to fix bugs around `block-scoped-var` rule.
